src/github.com/getlantern/flashlight/geolookup/geolookup.go

```diff
@@ -13,58 +17,92 @@ import (
 const (
 	messageType = `GeoLookup`
 
-	maxRetries = 10
+	basePublishInterval     = 30 * time.Second
+	publishIntervalVariance = basePublishInterval - 10*time.Second
 )
 
 var (
 	log = golog.LoggerFor("geolookup-flashlight")
 
-	service     *ui.Service
-	lookupMutex sync.Mutex
-	lookupData  *geolookup.City
+	service           *ui.Service
+	client            atomic.Value
+	lastKnownLocation atomic.Value
+	cfgMutex          sync.Mutex
 )
 
-func getUserGeolocationData(client *http.Client) (*geolookup.City, error) {
-	lookupMutex.Lock()
-	defer lookupMutex.Unlock()
+// Configure configures geolookup to use the given http.Client to perform
+// lookups. geolookup runs in a continuous loop, periodically updating its
+// location and publishing updates to any connected clients. We do this
+// continually in order to detect when the computer's location has changed.
+func Configure(newClient *http.Client) {
+	cfgMutex.Lock()
+	defer cfgMutex.Unlock()
 
-	if lookupData != nil {
-		// We already looked up IP's information.
-		return lookupData, nil
-	}
+	client.Store(newClient)
 
-	var err error
-	for i := 0; i < maxRetries; i++ {
-		// Will look up only if we're using a proxy.
-		lookupData, err = geolookup.LookupIPWithClient("", client)
-		if err == nil {
-			// We got what we wanted, no need to query for it again, let's exit.
-			return lookupData, nil
+	if service == nil {
+		err := registerService()
+		if err != nil {
+			log.Errorf("Unable to register service: %v", err)
+			return
 		}
+		go write()
+		go read()
+		log.Debug("Running")
 	}
-
-	return nil, fmt.Errorf("Unable to look up geolocation information in %d tries: %v", maxRetries, err)
 }
 
-// StartService initializes the geolocation websocket service using the given
-// http.Client to do the lookups
-func StartService(client *http.Client) (err error) {
+func registerService() error {
 	helloFn := func(write func(interface{}) error) error {
-		city, err := getUserGeolocationData(client)
-		if err != nil {
-			return err
+		location := lastKnownLocation.Load()
+		if location == nil {
+			log.Trace("No lastKnownLocation, not sending anything to client")
+			return nil
 		}
-
-		return write(city)
+		log.Trace("Sending last known location to new client")
+		return write(location)
 	}
 
-	if service, err = ui.Register(messageType, nil, helloFn); err != nil {
-		return fmt.Errorf("Unable to register channel: %q", err)
-	}
+	var err error
+	service, err = ui.Register(messageType, nil, helloFn)
+	return err
+}
 
-	go read()
+func write() {
```

Comment:
The logic here has changed.  We now have a continuous loop that's always checking the location, waiting a random amount in between.  We do this so that if the computer's location changes, we'll know about it and update the map.